configstack/module.go

```diff
@@ -52,7 +53,68 @@ func ResolveTerraformModules(terragruntConfigPaths []string, terragruntOptions *
 		return []*TerraformModule{}, err
 	}
 
-	return crosslinkDependencies(mergeMaps(modules, externalDependencies), canonicalTerragruntConfigPaths)
+	crossLinkedModules, err := crosslinkDependencies(mergeMaps(modules, externalDependencies), canonicalTerragruntConfigPaths)
+	if err != nil {
+		return []*TerraformModule{}, err
+	}
+
+	finalModules := removeExcludedDirs(crossLinkedModules, terragruntOptions)
+
+	return finalModules, nil
+}
+
+// removeExcludedDirs iterates over a module slice and removes all entries, which should be ignored via the terragrunt-exclude-dir CLI flag.
+// Modules are also removed if one of its dependencies are to be excluded. Returns the cleaned-up slice of modules.
+func removeExcludedDirs(modules []*TerraformModule, terragruntOptions *options.TerragruntOptions) []*TerraformModule {
+
+	// If no excludeDir is specified return the modules list instantly
+	if terragruntOptions.ExcludeDir == "" {
+		return modules
+	}
+
+	excludeGlobMatches := []string{}
+
+	// If possible, expand the glob to get all excluded filepaths
+	excludeGlobMatches, err := zglob.Glob(terragruntOptions.ExcludeDir)
+
+	if err != nil {
+		// Glob can not be expanded, using literal value
+		excludeGlobMatches = []string{terragruntOptions.ExcludeDir}
+	}
+
+	// If the expanded glob paths are relative, generate absolute paths
+	absoluteExludeDir := []string{}
+	for _, m := range excludeGlobMatches {
+		if strings.HasPrefix(m, "/") {
+			absoluteExludeDir = append(absoluteExludeDir, m)
+		} else {
+			absoluteExludeDir = append(absoluteExludeDir, fmt.Sprintf("%s/%s", terragruntOptions.WorkingDir, m))
```

Comment:
Use `filepath.Join` so this works across operating systems.

Reply:
In fact, replace the body of this entire loop with a call to Terragrunts `util.CanonicalPath`, as, based on my reading of your code below, that's what you actually need.

Reply:
Makes completely sense, please check the new code.